### ISSUE TITLE
Read out of bound in TGA files

### ISSUE THREAD

**Issue by me22bee** [libgd/libgd#697]
hello,
[this issue](https://github.com/libgd/libgd/issues/247#issue-163166532) is showing a read out of bound for a corrupted TGA [test.txt](https://github.com/libgd/libgd/files/6407187/test.txt) which is patched by adding some checks for ` gdGetBuf`.

although the patch prevents occurring this vulnerability I saw that this function (` gdGetBuf`) is used in `read_header_tga` too which there is no check for its return value again.

I changed the header of the file which was used for the previous CVE-2016-6132. In fact, I changed the first byte to `ff` which is assigned to `tga->identsize`. 
[file(it is a tga, not a really a txt) ](https://github.com/libgd/libgd/files/6407191/test.txt)

```
...
tga->identsize = header[0];
...
gdGetBuf(tga->ident, tga->identsize, ctx);
...
```

when I run the test with this input file ASAN shows this:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==12378==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7f685ae4a7a0 bp 0x7fff6bb8b740 sp 0x7fff6bb8b720 T0)
==12378==The signal is caused by a READ memory access.
==12378==Hint: address points to the zero page.
    #0 0x7f685ae4a79f in gdImageDestroy /tmp/libgd/src/gd.c:390
    #1 0x55c05b670432 in main /tmp/libgd/tests/tga/bug00084.c:11
    #2 0x7f685aa3709a in __libc_start_main ../csu/libc-start.c:308
    #3 0x55c05b670339 in _start (/tmp/libgd/build/Bin/test_tga_bug00084+0x2339)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /tmp/libgd/src/gd.c:390 in gdImageDestroy
==12378==ABORTING
```

Is it showing another vulnerability?

**Comment by me22bee** [comment]
any comment?

**Comment by cmb69** [comment]
There might be a bug there, but who needs to read TGAs in … checks calendar … 2021?

**Comment by pierrejoye** [comment]
@me22bee could you upload the full test case or do you use only bug00084.c?

**Comment by me22bee** [comment]
> do you use only bug00084.c?

yes.

**Comment by pierrejoye** [comment]
@me22bee could you attach the patched test file? Should be an easy fix. TGA is still used, indeed not for web but our targets go further than web devs :)

**Comment by me22bee** [comment]
> could you attach the patched test file?

I have sent and pull request and also attached the patch file here.
[0001-fix-read-out-of-bands-in-reading-tga-header-file.patch.txt](https://github.com/libgd/libgd/files/6838308/0001-fix-read-out-of-bands-in-reading-tga-header-file.patch.txt)

**Comment by pierrejoye** [comment]
Thank you @me22bee  :)

**Comment by me22bee** [comment]
your welcome,
Don't you assign a CVE number to it?
@pierrejoye

**Comment by pierrejoye** [comment]
let me check with the ***@***.***, we have access :)

**Comment by me22bee** [comment]
> let me check with

any update?

**Comment by carnil** [comment]
CVE-2021-38115 seems to have been assigned for this issue.

**Comment by mshah-aiondigital** [comment]
Will a release be provided with this security fix?

**Comment by oerdnj** [comment]
> Will a release be provided with this security fix?

FTR the fix has been released in 2.3.3

**Comment by cmb69** [comment]
For the record: this was not a security issue. The OP run bug00084.c against a modified bug00084.tga, which triggered a segfault. However, this was not due to any issues with reading the TGA, but the badly written bug00084.c, which has been [fixed in the meantime](https://github.com/libgd/libgd/commit/167ea1f4f0003f3e9f7ca1e586189e99cf33d47f#diff-14a20f0399b66f7abcafda3ecd8d7849bf587e7bf564c63d560ee3964177bf61); prior to that commit, bug00084.c did not check whether `gdImageCreateFromTga()` succeeded, but called `gdImageDestroy()` unconditionally. So if reading of the TGA file failed, it called `gdImageDestroy(NULL)` which results in the reported segfault.
